File: Typescript/src/complex.ts
const split = (s: string): string[] => {
    let realString = "0";
    let imagString = "0";
    let i = 1;
    for (; i < s.length; i += 1) {
        if (s.charAt(i) === "+" || s.charAt(i) === "-") {
            realString = s.substring(0, i);
            imagString = s.substring(i, s.length - 1);
            if (s.charAt(s.length - 1) !== "i") {
                throw new Error("can't parse complex");
            }
            break;
        }
        if (s.charAt(i) === "i") {
            if (i !== s.length - 1) {
                throw new Error("can;t parse complex");
            }
            imagString = s.substring(0, i);
            break;
        }
    }
    if (i === s.length) {
        realString = s;
    }

    return [realString, imagString];
};

export class Complex {
    public static readonly EPSILON = 0.00001;

    public static readonly add = (a: Complex, b: Complex) =>
        new Complex(a.real + b.real, a.imag + b.imag)

    public static readonly divide = (a: Complex, b: Complex): Complex => {
        const bConjugate = b.conjugate();
        const a1 = Complex.multiply(a, bConjugate);
        const b1 = Complex.multiply(b, bConjugate);

        return new Complex(a1.real / b1.real, a1.imag / b1.real);
    }

    public static readonly fromString = (s: string): Complex => {
        const strs = split(s);
        const real = Number.parseFloat(strs[0]);
        const imag = Number.parseFloat(strs[1]);
        if (Number.isNaN(real) || Number.isNaN(imag)) {
            throw new Error("can't parse complex");
        }

        return new Complex(real, imag);
    }

    public static readonly multiply = (a: Complex, b: Complex) =>
        new Complex(a.real * b.real - a.imag * b.imag, a.real * b.imag + a.imag * b.real)

    private imag: number;
    private real: number;
    public constructor(real = 0, imag = 0) {
        this.real = real;
        this.imag = imag;
    }

    public readonly conjugate = () => new Complex(this.real, -this.imag);

    public readonly equals = (o: Complex): boolean =>
        Math.abs(o.imag - this.imag) < Complex.EPSILON &&
        Math.abs(o.real - this.real) < Complex.EPSILON

    public readonly getImag = () => this.imag;

    public readonly getReal = () => this.real;

    public readonly toString = (): string => {
        const real = (Math.round(this.real * 10000) / 10000).toString();
        const imag = (Math.round(this.imag * 10000) / 10000).toString();

        if (Math.abs(this.imag) < Complex.EPSILON) {
            return real;
        }
        if (Math.abs(this.real) < Complex.EPSILON) {
            return `${imag}i`;
        }
        const prefix = (this.imag > 0) ? "+" : "";

        return `${real}${prefix}${imag}i`;
    }
}

File: DotNET/src/Complex.cs
// This is an independent project of an individual developer. Dear PVS-Studio, please check it.
// PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
using System;
namespace src
{
    public class Complex
    {
        public static readonly double EPSILON = 0.00001;

        public static bool operator ==(Complex a, Complex b)
        {
            return !(a is null) && !(b is null) && Math.Abs(a.Imag - b.Imag) < EPSILON && Math.Abs(a.Real - b.Real) < EPSILON;
        }

        public static bool operator !=(Complex a, Complex b)
        {
            return !(a == b);
        }

        public static Complex operator +(Complex a, Complex b)
        {
            return new Complex(a.Real + b.Real, a.Imag + b.Imag);
        }

        public static Complex operator /(Complex a, Complex b)
        {
            Complex bConjugate = b.Conjugate();
            Complex a1 = a * bConjugate;
            Complex b1 = b * bConjugate;

            return new Complex(a1.Real / b1.Real, a1.Imag / b1.Real);
        }

        public static Complex operator *(Complex a, Complex b)
        {
            return new Complex(a.Real * b.Real - a.Imag * b.Imag, a.Real * b.Imag + a.Imag * b.Real);
        }

        public double Real { get; private set; }
        public double Imag { get; private set; }

        public Complex(double real = 0.0, double imag = 0.0)
        {
            Real = real;
            Imag = imag;
        }

        public Complex(string s)
        {
            IFormatProvider iFormatProvider = new System.Globalization.CultureInfo("en-US");
            String[] strs = Split(s);
            Real = Double.Parse(strs[0], iFormatProvider);
            Imag = Double.Parse(strs[1], iFormatProvider);
        }

        public Complex Conjugate()
        {
            return new Complex(Real, -Imag);
        }

        public override string ToString()
        {
            IFormatProvider iFormatProvider = new System.Globalization.CultureInfo("en-US");
            if (Math.Abs(Imag) < EPSILON)
            {
                return string.Format(iFormatProvider, "{0:0.####}", Real);
            }
            if (Math.Abs(Real) < EPSILON)
            {
                return string.Format(iFormatProvider, "{0:0.####}i", Imag);
            }
            string prefix = (Imag > 0) ? "+" : "";
            return string.Format(iFormatProvider, "{0:0.####}{1}{2:0.####}i", Real, prefix, Imag);
        }

        private string[] Split(string s)
        {
            string realString = "0";
            string imagString = "0";
            int i = 1;
            for (; i < s.Length; ++i)
            {
                if (s[i] == '+' || s[i] == '-')
                {
                    realString = s.Substring(0, i);
                    imagString = s.Substring(i, s.Length - i - 1);
                    if (s[s.Length - 1] != 'i')
                    {
                        throw new FormatException("can't parse complex");
                    }
                    break;
                }
                if (s[i] == 'i')
                {
                    if (i != s.Length - 1)
                    {
                        throw new FormatException("can't parse complex");
                    }
                    imagString = s.Substring(0, i);
                    break;
                }
            }
            if (i == s.Length)
            {
                realString = s;
            }
            return new string[] { realString, imagString };
        }

        public override bool Equals(object obj)
        {
            return obj is Complex && this == (Complex)obj;
        }

        public override int GetHashCode()
        {
            return ToString().GetHashCode();
        }
    }
}
